File: uikit/src/main/java/com/sendbird/uikit/model/MessageListUIParams.java
package com.sendbird.uikit.model;

import androidx.annotation.NonNull;

import com.sendbird.uikit.consts.MessageGroupType;

/**
 * Describes a configuration of a message item view
 *
 * since 3.3.0
 */
final public class MessageListUIParams {
    @NonNull
    private final MessageGroupType messageGroupType;
    private final boolean useMessageGroupUI;
    private final boolean useReverseLayout;
    private final boolean useQuotedView;
    private final boolean useMessageReceipt;

    private MessageListUIParams(@NonNull MessageGroupType messageGroupType, boolean useMessageGroupUI, boolean useReverseLayout, boolean useQuotedView, boolean useMessageReceipt) {
        this.messageGroupType = messageGroupType;
        this.useMessageGroupUI = useMessageGroupUI;
        this.useReverseLayout = useReverseLayout;
        this.useQuotedView = useQuotedView;
        this.useMessageReceipt = useMessageReceipt;
    }

    /**
     * Returns the type of message group UI.
     *
     * @return The value of {@link MessageGroupType}
     * since 3.3.0
     */
    @NonNull
    public MessageGroupType getMessageGroupType() {
        return messageGroupType;
    }

    /**
     * Returns whether the quoted view is used.
     *
     * @return <code>true</code> if the quoted view is used, <code>false</code> otherwise
     * since 3.3.0
     */
    public boolean shouldUseQuotedView() {
        return useQuotedView;
    }

    /**
     * Returns whether the message grouping is used.
     *
     * @return <code>true</code> if the message grouping is used, <code>false</code> otherwise
     * since 3.3.0
     */
    public boolean shouldUseMessageGroupUI() {
        return useMessageGroupUI;
    }

    /**
     * Returns whether the message list is reversed.
     *
     * @return <code>true</code> if the message list is reversed, <code>false</code> otherwise
     * since 3.3.0
     */
    public boolean shouldUseReverseLayout() {
        return useReverseLayout;
    }

    /**
     * Returns whether the status (read receipt, delivery receipt) of messages is shown.
     *
     * @return <code>true</code> if the message receipt is shown, <code>false</code> otherwise
     * since 3.3.0
     */
    public boolean shouldUseMessageReceipt() {
        return useMessageReceipt;
    }

    public static class Builder {
        @NonNull
        private MessageGroupType messageGroupType = MessageGroupType.GROUPING_TYPE_SINGLE;
        private boolean useMessageGroupUI = true;
        private boolean useReverseLayout = true;
        private boolean useQuotedView = false;
        private boolean useMessageReceipt = true;

        /**
         * Constructor
         *
         * since 3.3.0
         */
        public Builder() {}

        /**
         * Constructor
         *
         * @param params The message draw parameter to be used as the base
         * since 3.3.0
         */
        public Builder(@NonNull MessageListUIParams params) {
            this.messageGroupType = params.messageGroupType;
            this.useMessageGroupUI = params.useMessageGroupUI;
            this.useReverseLayout = params.useReverseLayout;
            this.useQuotedView = params.useQuotedView;
            this.useMessageReceipt = params.useMessageReceipt;
        }

        /**
         * Sets the type of message group UI.
         *
         * @param messageGroupType The value of {@link MessageGroupType}
         * @return This Builder object to allow for chaining of calls to set methods.
         * since 3.3.0
         */
        @NonNull
        public Builder setMessageGroupType(@NonNull MessageGroupType messageGroupType) {
            this.messageGroupType = messageGroupType;
            return this;
        }

        /**
         * Sets whether the quoted view is used.
         *
         * @param useQuotedView <code>true</code> if the quoted view is used, <code>false</code> otherwise
         * @return This Builder object to allow for chaining of calls to set methods.
         * since 3.3.0
         */
        @NonNull
        public Builder setUseQuotedView(boolean useQuotedView) {
            this.useQuotedView = useQuotedView;
            return this;
        }

        /**
         * Sets whether the message grouping is used.
         *
         * @param useMessageGroupUI <code>true</code> if the message grouping is used, <code>false</code> otherwise
         * @return This Builder object to allow for chaining of calls to set methods.
         * since 3.3.0
         */
        @NonNull
        public Builder setUseMessageGroupUI(boolean useMessageGroupUI) {
            this.useMessageGroupUI = useMessageGroupUI;
            return this;
        }

        /**
         * Sets whether the message list is reversed.
         *
         * @param useReverseLayout <code>true</code> if the message list is reversed, <code>false</code> otherwise
         * @return This Builder object to allow for chaining of calls to set methods.
         * since 3.3.0
         */
        @NonNull
        public Builder setUseReverseLayout(boolean useReverseLayout) {
            this.useReverseLayout = useReverseLayout;
            return this;
        }

        /**
         * Sets whether the status (read receipt, delivery receipt) of messages is shown.
         *
         * @param useMessageReceipt <code>true</code> if the message receipt is shown, <code>false</code> otherwise
         * @return This Builder object to allow for chaining of calls to set methods.
         * since 3.3.0
         */
        @NonNull
        public Builder setUseMessageReceipt(boolean useMessageReceipt) {
            this.useMessageReceipt = useMessageReceipt;
            return this;
        }

        /**
         * Builds an {@link MessageListUIParams} with the properties supplied to this builder.
         *
         * @return The {@link MessageListUIParams} from this builder instance.
         * since 3.3.0
         */
        @NonNull
        public MessageListUIParams build() {
            return new MessageListUIParams(this.messageGroupType, this.useMessageGroupUI, this.useReverseLayout, this.useQuotedView, this.useMessageReceipt);
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        MessageListUIParams that = (MessageListUIParams) o;

        if (useMessageGroupUI != that.useMessageGroupUI) return false;
        if (useReverseLayout != that.useReverseLayout) return false;
        if (useQuotedView != that.useQuotedView) return false;
        if (useMessageReceipt != that.useMessageReceipt) return false;
        return messageGroupType == that.messageGroupType;
    }

    @Override
    public int hashCode() {
        int result = messageGroupType.hashCode();
        result = 31 * result + (useMessageGroupUI ? 1 : 0);
        result = 31 * result + (useReverseLayout ? 1 : 0);
        result = 31 * result + (useQuotedView ? 1 : 0);
        result = 31 * result + (useMessageReceipt ? 1 : 0);
        return result;
    }

    @Override
    @NonNull
    public String toString() {
        return "MessageDrawParams{" +
                "messageGroupType=" + messageGroupType +
                ", useMessageGroupUI=" + useMessageGroupUI +
                ", useReverseLayout=" + useReverseLayout +
                ", useQuotedView=" + useQuotedView +
                '}';
    }
}

File: uikit/src/main/java/com/sendbird/uikit/model/MessageListUIParams.kt
package com.sendbird.uikit.model

import com.sendbird.uikit.consts.MessageGroupType
import com.sendbird.uikit.model.configurations.ChannelConfig
import com.sendbird.uikit.model.configurations.OpenChannelConfig
import com.sendbird.uikit.model.configurations.UIKitConfig
import com.sendbird.uikit.model.configurations.UIKitConfig.groupChannelConfig

/**
 * Describes a configuration of a message item view
 * @since 3.3.0
 */
data class MessageListUIParams internal constructor(
    /**
     * Returns the type of message group UI.
     *
     * @return The value of [MessageGroupType]
     * @since 3.3.0
     */
    val messageGroupType: MessageGroupType,
    private val useMessageGroupUI: Boolean,
    private val useReverseLayout: Boolean,
    private val useQuotedView: Boolean,
    private val useMessageReceipt: Boolean,
    /**
     * Returns [ChannelConfig] that contains the configuration of the channel.
     *
     * @return The value of [ChannelConfig]
     * @since 3.6.0
     * @see ChannelConfig
     */
    val channelConfig: ChannelConfig,
    /**
     * Returns [OpenChannelConfig] that contains the configuration of the open channel.
     *
     * @return The value of [OpenChannelConfig]
     * @since 3.6.0
     * @see OpenChannelConfig
     */
    val openChannelConfig: OpenChannelConfig
) {

    /**
     * Returns whether the quoted view is used.
     *
     * @return `true` if the quoted view is used, `false` otherwise
     * @since 3.3.0
     */
    fun shouldUseQuotedView(): Boolean {
        return useQuotedView
    }

    /**
     * Returns whether the message grouping is used.
     *
     * @return `true` if the message grouping is used, `false` otherwise
     * @since 3.3.0
     */
    fun shouldUseMessageGroupUI(): Boolean {
        return useMessageGroupUI
    }

    /**
     * Returns whether the message list is reversed.
     *
     * @return `true` if the message list is reversed, `false` otherwise
     * @since 3.3.0
     */
    fun shouldUseReverseLayout(): Boolean {
        return useReverseLayout
    }

    /**
     * Returns whether the status (read receipt, delivery receipt) of messages is shown.
     *
     * @return `true` if the message receipt is shown, `false` otherwise
     * @since 3.3.0
     */
    fun shouldUseMessageReceipt(): Boolean {
        return useMessageReceipt
    }

    class Builder {
        private var messageGroupType = MessageGroupType.GROUPING_TYPE_SINGLE
        private var useMessageGroupUI = true
        private var useReverseLayout = true
        private var useQuotedView = false
        private var useMessageReceipt = true
        private var channelConfig = groupChannelConfig
        private var openChannelConfig = UIKitConfig.openChannelConfig

        /**
         * Constructor
         * @since 3.3.0
         */
        constructor()

        /**
         * Constructor
         *
         * @param params The message draw parameter to be used as the base
         * @since 3.3.0
         */
        constructor(params: MessageListUIParams) {
            messageGroupType = params.messageGroupType
            useMessageGroupUI = params.useMessageGroupUI
            useReverseLayout = params.useReverseLayout
            useQuotedView = params.useQuotedView
            useMessageReceipt = params.useMessageReceipt
            channelConfig = params.channelConfig
            this.openChannelConfig = params.openChannelConfig
        }

        /**
         * Sets the type of message group UI.
         *
         * @param messageGroupType The value of [MessageGroupType]
         * @return This Builder object to allow for chaining of calls to set methods.
         * @since 3.3.0
         */
        fun setMessageGroupType(messageGroupType: MessageGroupType): Builder {
            this.messageGroupType = messageGroupType
            return this
        }

        /**
         * Sets whether the quoted view is used.
         *
         * @param useQuotedView `true` if the quoted view is used, `false` otherwise
         * @return This Builder object to allow for chaining of calls to set methods.
         * @since 3.3.0
         */
        fun setUseQuotedView(useQuotedView: Boolean): Builder {
            this.useQuotedView = useQuotedView
            return this
        }

        /**
         * Sets whether the message grouping is used.
         *
         * @param useMessageGroupUI `true` if the message grouping is used, `false` otherwise
         * @return This Builder object to allow for chaining of calls to set methods.
         * @since 3.3.0
         */
        fun setUseMessageGroupUI(useMessageGroupUI: Boolean): Builder {
            this.useMessageGroupUI = useMessageGroupUI
            return this
        }

        /**
         * Sets whether the message list is reversed.
         *
         * @param useReverseLayout `true` if the message list is reversed, `false` otherwise
         * @return This Builder object to allow for chaining of calls to set methods.
         * @since 3.3.0
         */
        fun setUseReverseLayout(useReverseLayout: Boolean): Builder {
            this.useReverseLayout = useReverseLayout
            return this
        }

        /**
         * Sets whether the status (read receipt, delivery receipt) of messages is shown.
         *
         * @param useMessageReceipt `true` if the message receipt is shown, `false` otherwise
         * @return This Builder object to allow for chaining of calls to set methods.
         * @since 3.3.0
         */
        fun setUseMessageReceipt(useMessageReceipt: Boolean): Builder {
            this.useMessageReceipt = useMessageReceipt
            return this
        }

        /**
         * Sets [ChannelConfig] that contains the configuration of the channel.
         * Use `UIKitConfig.groupChannelConfig.clone()` for the default value.
         * Example usage:
         *
         * ```kotlin
         * val params = MessageListUIParams.Builder()
         *                 .setChannelConfig(UIKitConfig.groupChannelConfig.clone().apply {
         *                     this.enableMention = true
         *                 }).build()
         * ```
         * @param channelConfig The value of [ChannelConfig]
         * @return This Builder object to allow for chaining of calls to set methods.
         * @since 3.6.0
         * @see ChannelConfig
         */
        fun setChannelConfig(channelConfig: ChannelConfig): Builder {
            this.channelConfig = channelConfig
            return this
        }

        /**
         * Sets [OpenChannelConfig] that contains the configuration of the open channel.
         * Use `UIKitConfig.openChannelConfig.clone()` for the default value.
         * Example usage:
         *
         *```kotlin
         * val params = MessageListUIParams.Builder()
         *             .setOpenChannelConfig(UIKitConfig.openChannelConfig.clone().apply {
         *                  this.enableOgTag = false
         *             }).build()
         * ```
         *
         *
         * @param openChannelConfig The value of [OpenChannelConfig]
         * @return This Builder object to allow for chaining of calls to set methods.
         * @since 3.6.0
         * @see OpenChannelConfig
         */
        fun setOpenChannelConfig(openChannelConfig: OpenChannelConfig): Builder {
            this.openChannelConfig = openChannelConfig
            return this
        }

        /**
         * Builds an [MessageListUIParams] with the properties supplied to this builder.
         *
         * @return The [MessageListUIParams] from this builder instance.
         * @since 3.3.0
         */
        fun build(): MessageListUIParams {
            return MessageListUIParams(
                messageGroupType,
                useMessageGroupUI,
                useReverseLayout,
                useQuotedView,
                useMessageReceipt,
                channelConfig,
                this.openChannelConfig
            )
        }
    }
}
